KAFKA-8449: Restart tasks on reconfiguration under incremental cooperative rebalancing

kkonstantine opened the pull request:
With incremental cooperative rebalancing tasks are not restarted unless they are reassigned to different worker. Therefore, to load config updates due to connector reconfiguration the currently running tasks that get new configs need to be recorded and considered for restart when connectors and tasks are started after a rebalance. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kkonstantine committed:
KAFKA-8449: Restart task on reconfiguration under incremental cooperative rebalancing

kkonstantine commented:
@rhauch @ewencp this PR aims to fix the issue with tasks that don't pick up new configs under reconfiguration. Your thoughts are appreciated

kkonstantine commented:
jk8 tests green. jdk11 has one failure in a not relevant test case (`kafka.network.DynamicConnectionQuotaTest.testDynamicConnectionQuota`).

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -900,6 +995,8 @@ private void startWork() {
         }
 
+        runningAssignment.tasks().removeAll(tasksToRestart);
+        tasksToRestart.clear();
```
Minor: it would help to have a comment here to explain why the tasks to restart are removed from the running tasks.

kkonstantine replied:
Done

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -360,6 +346,100 @@ public void tick() {
+
+            if (needsReconfigRebalance) {
+                member.requestRejoin();
+                log.debug("Requesting rebalance due to reconfiguration of tasks (needsReconfigRebalance: {})",
```
Minor: this log line is in a different location in the similar `updateConfigsWithEager(...)` method. Would it make sense to move this log message to just before the `member.requestRejoin()` call?

kkonstantine replied:
Moved it to both before `member.requestRejoin`. I changed it because it was even after we set this variable to false, which made the log message a bit confusing.

kkonstantine replied:
(done)

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java:
```diff
@@ -360,6 +346,100 @@ public void tick() {
+                // This set is unused in this policy, but emptied here to avoid branching on
+                // protocol version in the onTaskConfigUpdate callback
+                taskConfigUpdates.clear();
+                log.debug("Requesting rebalance due to reconfiguration of tasks (needsReconfigRebalance: {})",
```
Would it make sense to move this log message to just before the `member.requestRejoin()` call?

kkonstantine replied:
Done

rhauch reviewed (commented):
Thanks, @kkonstantine! These changes look good, but I have a few pretty minor suggestions.

kkonstantine committed:
KAFKA-8449: Keep execution paths separate for config updates between eager and cooperative

Also include the group generation in the log message when the worker receives its assignment

kkonstantine committed:
KAFKA-8449: Add comments and improve logging

kkonstantine commented:
Thanks @rhauch ! 
I addressed your comments and pushed an update that keeps the two execution paths separate between eager and cooperative. Waiting for the unit tests and will run a round of system tests too.

rhauch reviewed (approved):
LGTM. Thanks, @kkonstantine!
